src/System.Diagnostics.Debug/src/System/Diagnostics/Debug.Windows.cs

```diff
@@ -0,0 +1,132 @@
+
+        // --------------
+        // PAL ENDS HERE
+        // --------------
+
+        internal class WindowsDebugLogger : IDebugLogger
+        {
+            [SecuritySafeCritical]
+            public void ShowAssertDialog(string stackTrace, string message, string detailMessage)
+            {
+                string fullMessage = message + Environment.NewLine + detailMessage + Environment.NewLine + stackTrace;
+
+                int flags = Interop.User32.MB_OKCANCEL | Interop.User32.MB_ICONHAND | Interop.User32.MB_TOPMOST;
+
+                if (IsRTLResources)
+                {
+                    flags = flags | Interop.User32.MB_RIGHT | Interop.User32.MB_RTLREADING;
+                }
+
+                int rval = 0;
+
+                // Run the message box on its own thread.
+                rval = new MessageBoxPopup(fullMessage, SR.DebugAssertTitle, flags).ShowMessageBox();
+
+                switch (rval)
+                {
+                    case Interop.User32.IDCANCEL:
+                        if (!System.Diagnostics.Debugger.IsAttached)
+                        {
+                            System.Diagnostics.Debugger.Launch();
+                        }
+                        System.Diagnostics.Debugger.Break();
+                        break;
+                    default:
+                        Debug.Assert(rval == Interop.User32.IDOK);
+                        break;
+                }
+            }
+
+            public void WriteLineCore(string message)
+            {
+                message = message + "\r\n";
+                Write(message);
+            }
+
+            public void WriteCore(string message)
+            {
+                // really huge messages mess up both VS and dbmon, so we chop it up into 
+                // reasonable chunks if it's too big. This is the number of characters 
+                // that OutputDebugstring chunks at.
+                const int WriteChunkLength = 4091;
+
+                // We don't want output from multiple threads to be interleaved.
+                lock (s_ForLock)
+                {
+                    if (message == null || message.Length <= WriteChunkLength)
+                    {
+                        WriteToDebugger(message);
+                    }
+                    else
+                    {
+                        int offset;
+                        for (offset = 0; offset < message.Length - WriteChunkLength; offset += WriteChunkLength)
+                        {
+                            WriteToDebugger(message.Substring(offset, WriteChunkLength));
+                        }
+                        WriteToDebugger(message.Substring(offset));
+                    }
+                }
+            }
+
+            [System.Security.SecuritySafeCritical]
+            private static void WriteToDebugger(string message)
+            {
+                Interop.mincore.OutputDebugString(message ?? string.Empty);
```

Comment:
If message is null, should we instead skip calling OutputDebugString altogether, or is there a good reason to still call it with the empty string?